lightly/models/modules/sparse/spark.py

```diff
@@ -0,0 +1,258 @@
+    The ordering of the dimensions in the inputs. channels_last corresponds to inputs with
+    shape (batch_size, height, width, channels) while channels_first corresponds to inputs
+    with shape (batch_size, channels, height, width).
+
+
+    Attributes:
+        normalized_shape:
+            Input shape from an expected input of size
+            normalized_shape or a single integer.
+        eps:
+            A value added to the denominator for numerical stability. Default: 1e-6.
+        data_format:
+            The ordering of the dimensions in the inputs. Default: "channels_last".
+        sparse:
+            Whether to use sparse computation. Default: True.
+
+    """
+
+    def __init__(
+        self,
+        normalized_shape: int,
+        eps: float = 1e-6,
+        data_format: Literal["channels_last", "channels_first"] = "channels_last",
+        sparse: bool = True,
+    ):
+        if data_format not in ["channels_last", "channels_first"]:
+            raise NotImplementedError
+        super().__init__(normalized_shape, eps, elementwise_affine=True)
+        self.data_format = data_format
+        self.sparse = sparse
+
+    def forward(self, x):
+        if x.ndim == 4:
+            if self.data_format == "channels_last":
+                if self.sparse:
+                    ii = _get_active_ex_or_ii(
+                        H=x.shape[1], W=x.shape[2], returning_active_ex=False
+                    )
+                    nc = x[ii]
+                    nc = super(SparseConvNeXtLayerNorm, self).forward(nc)
+
+                    x = torch.zeros_like(x)
+                    x[ii] = nc
+                    return x
+                else:
+                    return super(SparseConvNeXtLayerNorm, self).forward(x)
+            else: 
+                if self.sparse:
+                    ii = _get_active_ex_or_ii(
+                        H=x.shape[2], W=x.shape[3], returning_active_ex=False
+                    )
+                    bhwc = x.permute(0, 2, 3, 1)
+                    nc = bhwc[ii]
+                    nc = super(SparseConvNeXtLayerNorm, self).forward(nc)
+
+                    x = torch.zeros_like(bhwc)
+                    x[ii] = nc
+                    return x.permute(0, 3, 1, 2)
+                else:
+                    u = x.mean(1, keepdim=True)
+                    s = (x - u).pow(2).mean(1, keepdim=True)
+                    x = (x - u) / torch.sqrt(s + self.eps)
+                    x = self.weight[:, None, None] * x + self.bias[:, None, None]
+                    return x
+        else:  # BLC or BC
+            if self.sparse:
+                raise NotImplementedError
+            else:
+                return super(SparseConvNeXtLayerNorm, self).forward(x)
+
+    def __repr__(self):
+        return (
+            super(SparseConvNeXtLayerNorm, self).__repr__()[:-1]
+            + f', ch={self.data_format.split("_")[-1]}, sp={self.sparse})'
+        )
+
+
+class SparseConvNeXtBlock(nn.Module):
```

Comment:
What is the difference between this ConvNeXt implementation and the ones from torchvision and timm? Would it be possible to inherit from torchvisions `CNBlock` instead as it is used in `dense_model_to_sparse`? This way we wouldn't have to implement `DropPath` ourselves.